feat: plugin implementation

kbudde opened the pull request:
first example of plugin integration

It will list plugins on running `myks`
```
❯ myks
...bash
Usage:
  myks [command]

Plugin Subcommands:
  cmd1         Execute cmd1
  cmd2         Execute cmd2

Additional Commands:
  all          Run sync and render
  completion   Generate the autocompletion script for the specified shell
  help         Help about any command
  init         Initialize new myks project
  plugin       Plugin commands
  print-config Print myks configuration
  render       Render manifests
  sync         Sync vendir configs
...
```

Plugin is called once for each application. Smartmode is supported.

```bash
❯ myks cmd2
10:39PM INF Setting log level to: debug
10:39PM INF Using config file: /home/kbudde/projects/private/lab/.myks.yaml
10:39PM DBG Running git to: set git path prefix
git -C . rev-parse --show-prefix
10:39PM DBG Running git to: set git repository url
git remote get-url origin
...
[cmd2] zoo/goldpinger: executing...
[cmd2] zoo/goldpinger: map[string]string{"MYKS_APP":"goldpinger", "MYKS_APP_PROTOTYPE":"prototypes/goldpinger", "MYKS_ARGOCD_ENABLED":"false", "MYKS_ENV":"zoo", "MYKS_ENV_DATA_FILE":"envs/zoo/.myks/tmp/env-data.yaml", "MYKS_ENV_DIR":"envs/zoo", "MYKS_RENDERED_APP_DIR":"rendered/envs/zoo/goldpinger"}

```

kbudde commented:
@fritzduchardt , @Zebradil  This PR adds generic plugins to myks, e.g. `myks-apply`.
Plugin is called once for each application.

fritzduchardt commented:
Nice work! One idea regarding the placement of plugins for myks to pick them up. I would like to place them in a dedicated global folder, e.g. ~/myks/plugins. In this case prefixing them should not be necessary. Also, with my setup, the iteration over all path entries leads to a bunch of errors, since I don't clean that up regularly and have non-existing folders in there. Alternatively, plugin registration could also be part of the root env-data.yaml. That way binaries don't need to get duplicated. 

A couple of ideas and questions regarding the execution:
- I suppose this will really become useful, once we can pass arguments to the plugin. How do you plan to go about this? Maybe something like `myks kapp envs/env1 ALL -- deploy -a $MYKS_APP -f $MYKS_RENDERED_APP_DIR`
- If we have a means to register plugins in the env-data.yaml, we could also add default arguments there. This would also open the possibility to hard-wire plugin execution straight after rendering, e.g. for kubeconform.

kbudde commented:
Good points, passing arguments to plugins makes probably sense from the beginning. Your proposal looks good, will do :) 

For the registering of plugins: I would add another PR to add "Hooks" to myks.  Not sure, how it will look like but I think it's related to plugins somehow...

About the prefix and env: I was using the same logic as kubectl. I changed the error level to trace. To hide it more :) 
What do we need here?
Right now, it loads plugins from env folders and from "plugins" folder in the current repository. I think, it's a more common way of handling plugins. But for sure, it can be changed.

Zebradil commented:
Don't forget that we have the `.myks.yaml` where plugin sources can be defined. The sources can be organized in a desired way.

Zebradil commented on internal/myks/plugins.go:
```diff
@@ -0,0 +1,136 @@
+}
+
+func (p PluginCmd) Exec(a *Application) error {
+	env := map[string]string{
+		"MYKS_ENV":              a.e.Id,
+		"MYKS_APP":              a.Name,
+		"MYKS_APP_PROTOTYPE":    a.Prototype,
+		"MYKS_ENV_DIR":          a.e.Dir,
+		"MYKS_ENV_DATA_FILE":    a.e.renderedEnvDataFilePath,
+		"MYKS_RENDERED_APP_DIR": "rendered/envs/" + a.e.Id + "/" + a.Name, // TODO: provide func and use it everywhere,
+		"MYKS_ARGOCD_ENABLED":   strconv.FormatBool(a.argoCDEnabled),
+	}
+	if a.argoCDEnabled {
+		env["MYKS_ARGOCD_APP_NAME"] = a.getArgoCDDestinationDir()
+		env["MYKS_ARGOCD_APP_PROJECT"] = "rendered/argocd/" + a.e.Id + "/" + a.Name // TODO: provide func and use it everywhere,
+	}
```
We should consider a way to pass data/values of the current app. Something like this, maybe?

```go
env["MYKS_DATA_VALUES"] = a.ytt(p.Name(), "get data values", h.app.yttDataFiles, "--data-values-inspect")
```

kbudde commented:
> Don't forget that we have the `.myks.yaml` where plugin sources can be defined. The sources can be organized in a desired way.

OK, sounds like we define a list of folders with myks plugins in .myks.yaml. The "myks-" prefix is not needed in this case, correct?

Zebradil commented:
> The "myks-" prefix is not needed in this case, correct?

It depends. I'd do the following:

- if a plugin is discovered via the `PATH`, it should have the prefix,
- if a plugin is discovered via the config file, the prefix is not needed.

At the moment, I don't think discovery via the `PATH` is valuable. But I'm not entirely sure about that.

kbudde commented:
Using the config file is harder than expected.

At the moment, the config file is linked with `rootCmd.PersistentFlags()`. The config file itself is initialized in `cobra.OnInitialize(initConfig)`. At this time I can not add any additional commands to the root command which is required to make plugins work.

I see only two options, if we want to make the plugin sources configurable: 

1. easy: no myks flag and plugin.init() is reading the .myks.yaml to get the sources directly
2. hard: drop all the init functions and configure cobra command loading/init in our main function. This should allow us to load config earlier.

I would go for 1.. 2. can be done later.

kbudde commented:
Ok, I dropped the init functions and changed how we initialize the root cmd; it was easier 🤦🏽 

The PR is now a bit bigger but you can now:
-load plugins from PATH, using the prefix `myks-`
- Configure plugin sources in .myks.yaml. Executables in these folders are not using the prefix
	- Defaults to `["./plugins"]`. Maybe we extend the default to include as well a global folder (e.g. `$HOME/myks/plugins`) as suggested by @fritzduchardt 
	- Plugin names are unique, last one wins. 
- I removed the dedicated "plugin" cmd as it was only listing plugins. Just calling `myks` will list the plugins as well.

Zebradil commented on internal/myks/plugins.go:
```diff
@@ -0,0 +1,150 @@
+}
+
+func (p PluginCmd) generateEnv(a *Application) (map[string]string, error) {
+	env := map[string]string{
+		"MYKS_ENV":              a.e.Id,
+		"MYKS_APP":              a.Name,
+		"MYKS_APP_PROTOTYPE":    a.Prototype,
+		"MYKS_ENV_DIR":          a.e.Dir,
+		"MYKS_ENV_DATA_FILE":    a.e.renderedEnvDataFilePath,
+		"MYKS_RENDERED_APP_DIR": "rendered/envs/" + a.e.Id + "/" + a.Name, // TODO: provide func and use it everywhere,
+		"MYKS_ARGOCD_ENABLED":   strconv.FormatBool(a.argoCDEnabled),
+	}
+	if a.argoCDEnabled {
+		env["MYKS_ARGOCD_APP_NAME"] = a.getArgoCDDestinationDir()
+		env["MYKS_ARGOCD_APP_PROJECT"] = "rendered/argocd/" + a.e.Id + "/" + a.Name // TODO: provide func and use it everywhere,
+	}
```
```suggestion
	env := map[string]string{
		"MYKS_ENV":              a.e.Id,
		"MYKS_APP":              a.Name,
		"MYKS_APP_PROTOTYPE":    a.Prototype,
		"MYKS_ENV_DIR":          a.e.Dir,
		"MYKS_RENDERED_APP_DIR": "rendered/envs/" + a.e.Id + "/" + a.Name, // TODO: provide func and use it everywhere
	}
```

Since we now have `MYKS_DATA_VALUES`, most of the info can be retrieved from it. Anything hard to retrieve directly or impossible to retrieve at the moment (like the rendered dir, because it requires knowing the configuration of the `Globe` object), we can add into the data schema as we did with the `context` section. But I think it's convenient to have the most basic things as environment variables. So, I propose to remove ArgoCD-related stuff and `MYKS_ENV_DATA_FILE`, because it probably won't be used much.

Zebradil reviewed (approved):
Great work! LGTM, only one small suggestion.

fritzduchardt commented:
Would still be 

> Ok, I dropped the init functions and changed how we initialize the root cmd; it was easier 🤦🏽
> 
> The PR is now a bit bigger but you can now:
> 
> * load plugins from PATH, using the prefix `myks-`
> * Configure plugin sources in .myks.yaml. Executables in these folders are not using the prefix
>   
>   * Defaults to `["./plugins"]`. Maybe we extend the default to include as well a global folder (e.g. `$HOME/myks/plugins`) as suggested by @fritzduchardt
>   * Plugin names are unique, last one wins.
> * I removed the dedicated "plugin" cmd as it was only listing plugins. Just calling `myks` will list the plugins as well.

I did not get the initialization problem. If "plugin-sources" can be placed in in the .myks.yaml then a more detailed information about plugins should also fit there.

Zebradil commented:
@fritzduchardt could you provide an example of "more detailed information about plugins"? What else is missing in this PR?

fritzduchardt commented:
> @fritzduchardt could you provide an example of "more detailed information about plugins"? What else is missing in this PR?

It does not have to be part of this PR, but in general for this feature, I would like the following:
- Possibility to use current binaries, e.g. kubectl, kapp or kubeconfirm. 
- Central configuration for all gitops repos at user level.
- Default arguments for all binaries

One way to achieve this would be a new section in the .myks.yaml like this:
```
plugins:
    - name: kubectl
      path: /usr/bin/kubectl
      defaultArgs: apply -f $MYKS_RENDERED_APP_DIR
    - name: kubeconform
      path: /usr/bin/kubeconform
      defaultArgs: $MYKS_RENDERED_APP_DIR
```
Also, we should consider introducing a user level .myks.yaml, e.g. at ~/.config/myks/myks-config.yaml

Zebradil commented:
I think the most versatile way to achieve this is to use shell scripts to wrap the binaries.

Consider the following `.myks.yaml`:

```yaml
plugin-sources:
  # global collection of plugins
  - ~/.config/myks/plugins
  # plugins shipped with the project
  - ./plugins
```

Then, in `~/.config/myks/plugins/kapp`:

```bash
#!/bin/bash

kapp deploy --app "$MYKS_APP" --namespace kapp-store --file "$MYKS_RENDERED_APP_DIR" "$@"
```

This is a simplified example. More things can be done using `yq` and data values from `MYKS_DATA_VALUES` env var.

fritzduchardt commented:
> I think the most versatile way to achieve this is to use shell scripts to wrap the binaries.
> 
> Consider the following `.myks.yaml`:
> 
> ```yaml
> plugin-sources:
>   # global collection of plugins
>   - ~/.config/myks/plugins
>   # plugins shipped with the project
>   - ./plugins
> ```
> 
> Then, in `~/.config/myks/plugins/kapp`:
> 
> ```shell
> #!/bin/bash
> 
> kapp deploy --app "$MYKS_APP" --namespace kapp-store --file "$MYKS_RENDERED_APP_DIR" "$@"
> ```
> 
> This is a simplified example. More things can be done using `yq` and data values from `MYKS_DATA_VALUES` env var.

This way is certainly very powerful. Still, it requires more setup for the user. An inline solution in the .myks.yaml would be easier to set up. Also, we could provide commented out defaults that enable our users to quickly enable popular plugins. The more versatile script option would still work for pros.

Zebradil commented:
The more features we add to the config, the more time we spend on maintenance. Things that are set in schemas are hard to change later — this requires a deprecation period, warning messages, etc. I want to make sure that those convenience features are really needed before committing to them.

Instead, we can combine the two approaches:

```yaml
plugins:
  - name: kubectl
    cmd: |
      #!/bin/bash
      /usr/bin/kubectl apply -f $MYKS_RENDERED_APP_DIR
  - name: kubeconform
    cmd: |
      #!/bin/bash
      /usr/bin/kubeconform $MYKS_RENDERED_APP_DIR
```

Or something like that.

Zebradil commented:
I'm going to merge this one for now. No matter which way we decide to go, this can be done as a follow-up PR.
